### PR TITLE
Upload SimMIM model.

### PR DESCRIPTION
**Pull request by Ra1Nik**
Upload SimMIM model. It is a very similar architecture to MAE, but it uses a ViT encoder using as input both masked and non-masked patches, it has just a simple linear as a decoder and uses L1 instead of L2 loss.

Added Examples, utils function and test and documentation.

### CODE REVIEW

**Comment by Ra1Nik** [comment]
Rendering of docs
[SimMIM — lightly 1.2.36 documentation.pdf](https://github.com/lightly-ai/lightly/files/10175151/SimMIM.lightly.1.2.36.documentation.pdf)

File: docs/source/examples/simmim.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
+.. _simmim:
+
+SimMIM
+======
+
+Example implementation of SimMIM: A Simple Framework for Masked Image Modeling architecture. SimMIM is a
+very similar architecture to `Masked Autoencoders Are Scalable Vision Learners, 2021 <https://arxiv.org/abs/2111.06377>`_.
+It uses a ViT encoder using as input both masked and non-masked patches. Other differences with respect to MAE
+is that it has just a simple linear as a decoder and uses L1 instead of L2 loss.
```
Comment (guarin) [inline docs/source/examples/simmim.rst:9]: a simple linear as -> a simple linear layer

File: examples/pytorch/simmim.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import torch
+from torch import nn
+import torchvision
+
+from lightly.data import LightlyDataset
+from lightly.data.collate import MAECollateFunction #Same collate as MAE
```
Comment (guarin) [inline examples/pytorch/simmim.py:6]: Could you add a space after the # ?
Like this: `# Same collate as MAE`

File: examples/pytorch/simmim.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        # same backbone as MAE
+        self.backbone = masked_autoencoder.MAEBackbone.from_vit(vit) 
+
+        # the decoder is a simple linear layer
+        self.decoder = nn.Linear(vit.hidden_dim, vit.patch_size ** 2 * 3)
+        
+
+    def forward_encoder(self, images, batch_size, idx_keep=None):
+        # pass all the tokens to the encoder, both masked and non masked ones
+        tokens = self.backbone.images_to_tokens(images, prepend_class_token = True)
```
Comment (guarin) [inline examples/pytorch/simmim.py:30]: Could you remove the spaces here: `prepend_class_token = True` -> `prepend_class_token=True`?

File: lightly/models/utils.py [hunk 4]
```diff
@@ -347,6 +347,28 @@ def set_at_index(
     index = expand_index_like(index, tokens)
     return torch.scatter(tokens, 1, index, value)
 
+def mask_at_index(tokens, idx_keep, value):
+    """Copies all values into the input tensor at the given indices using mask instead
+    of scatter. Use this istead of set_at_index if you are concatenating learnable and
```
Comment (guarin) [inline lightly/models/utils.py:352]: istead -> instead

File: lightly/models/utils.py [hunk 5]
```diff
@@ -347,6 +347,28 @@ def set_at_index(
     index = expand_index_like(index, tokens)
     return torch.scatter(tokens, 1, index, value)
 
+def mask_at_index(tokens, idx_keep, value):
```
Comment (guarin) [inline lightly/models/utils.py:350]: Could you add types here:
```
def mask_at_index(tokens: torch.Tensor, index: torch.Tensor, value: torch.Tensor) -> torch.Tensor:
```

Comment (guarin) [reply lightly/models/utils.py:350]: And rename `idx_keep` to `index` as in `set_at_index`

File: tests/models/test_ModelUtils.py [hunk 6]
```diff
@@ -133,6 +133,19 @@ def test_set_at_index(self, seed=0):
             for j in range(index.shape[1]):
                 self._assert_tensor_equal(new_tokens[i, index[i, j]], values[i, j])
 
+    def test_mask_at_index(self, seed=0):
+        torch.manual_seed(seed)
+        index = torch.Tensor([
+            [1, 0, 3, 2],
+            [1, 2, 0, 3],
```
Comment (guarin) [inline tests/models/test_ModelUtils.py:140]: Could you make the index have dimension (2, 3) instead of (2, 4)? It should work if index does not have the same shape as tokens.

File: tests/models/test_ModelUtils.py [hunk 7]
```diff
@@ -133,6 +133,19 @@ def test_set_at_index(self, seed=0):
...
+        ]).long()
+        tokens = torch.rand(2, 4, 5)
+        values = torch.rand(2, 4, 5)
+        new_tokens = utils.mask_at_index(tokens, index, values)
+        new_tokens_set = utils.set_at_index(tokens, index, values)
+        # make sure that values are copied correctly
+        self._assert_tensor_equal(torch.sort(new_tokens, 1)[0], torch.sort(new_tokens_set,1)[0])
```
Comment (guarin) [inline tests/models/test_ModelUtils.py:147]: Shouldn't `self._assert_tensor_equal(new_tokens, new_tokens_set)` work as well?

**Review by guarin — COMMENTED** [review]
Awesome stuff!

I left some comments for minor improvements but we can merge this soon.

File: examples/pytorch/simmim.py (hunk 2 above)
Comment (guarin) [reply examples/pytorch/simmim.py:6]: Could you change this also for the other 2 versions (lightning, distributed)?

File: examples/pytorch/simmim.py (hunk 3 above)
Comment (guarin) [reply examples/pytorch/simmim.py:30]: Could you change this also for the other 2 versions (lightning, distributed)?

File: tests/models/test_ModelUtils.py (hunk 6 above)
Comment (Ra1Nik) [reply tests/models/test_ModelUtils.py:140]: I don't think so, because this mask is used just for equal dimension matrixes as input, right?

File: examples/pytorch/simmim.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    def forward(self, images):
+        batch_size = images.shape[0]
+        idx_keep, idx_mask = utils.random_token_mask(
+            size=(batch_size, self.sequence_length),
+            mask_ratio=self.mask_ratio,
+            device=images.device,
+        )
+        
+        # Encoding...
+        x_encoded = self.forward_encoder(images, batch_size, idx_keep)
```
Comment (guarin) [inline examples/pytorch/simmim.py:46]: Here we should pass `idx_mask` instead of `idx_keep`.

File: examples/pytorch/simmim.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        self.mask_token = nn.Parameter(torch.zeros(1, 1, decoder_dim))
+
+        # same backbone as MAE
+        self.backbone = masked_autoencoder.MAEBackbone.from_vit(vit) 
+
+        # the decoder is a simple linear layer
+        self.decoder = nn.Linear(vit.hidden_dim, vit.patch_size ** 2 * 3)
+        
+
+    def forward_encoder(self, images, batch_size, idx_keep=None):
```
Comment (guarin) [inline examples/pytorch/simmim.py:28]: Should be `idx_mask` instead of `idx_keep`. We should also remove the default `None` value as `mask_at_index` does not accept `None` as index.

File: examples/pytorch/simmim.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        # pass all the tokens to the encoder, both masked and non masked ones
+        tokens = self.backbone.images_to_tokens(images, prepend_class_token=True)
+        tokens_masked = utils.mask_at_index(tokens, idx_keep , self.mask_token)
```
Comment (guarin) [inline examples/pytorch/simmim.py:31]: Should pass `idx_mask` instead of `idx_keep`.

File: lightly/models/utils.py [hunk 11]
```diff
@@ -347,6 +347,30 @@ def set_at_index(
...
+) -> torch.Tensor:
+    """Copies mask token into the input tensor at the given indices.
+    
+    Args:
+        tokens:
+            Tokens tensor with shape (batch_size, sequence_length, dim).
+        index:
+            Index tensor with shape (batch_size, index_length).
+        value:
+            Value tensor with shape (1, 1, dim).
```
Comment (guarin) [inline lightly/models/utils.py:363]: Should be `mask_token` instead of `value`.

**Review by guarin — COMMENTED** [review]
Awesome! Just needs some final cleanup and then we should be good.

**Review by guarin — APPROVED** [review]
💯
